tempesta_fw/classifier/frang.c

```diff
@@ -0,0 +1,695 @@
+
+	frang_cfg.http_ct_vals = vals;
+	return 0;
+}
+
+static void
+frang_free_ct_vals(TfwCfgSpec *cs)
+{
+	kfree(frang_cfg.http_ct_vals);
+	frang_cfg.http_ct_vals = NULL;
+}
+
+static TfwCfgSpec frang_cfg_section_specs[] = {
+	{
+		"request_rate", "0",
+		tfw_cfg_set_int,
+		&frang_cfg.req_rate,
+	},
+	{
+		"request_burst", "0",
+		tfw_cfg_set_int,
+		&frang_cfg.req_burst,
+	},
+	{
+		"new_connection_rate", "0",
+		tfw_cfg_set_int,
+		&frang_cfg.conn_rate,
+	},
+	{
+		"new_connection_burst", "0",
+		tfw_cfg_set_int,
+		&frang_cfg.conn_burst,
+	},
+	{
+		"concurrent_connections", "0",
+		tfw_cfg_set_int,
+		&frang_cfg.conn_max,
+	},
+	{
+		"http_uri_len", "0",
+		tfw_cfg_set_int,
+		&frang_cfg.http_uri_len,
+	},
+	{
+		"http_field_len", "0",
+		tfw_cfg_set_int,
+		&frang_cfg.http_field_len,
+	},
+	{
+		"http_body_len", "0",
+		tfw_cfg_set_int,
+		&frang_cfg.http_body_len,
+	},
+	{
+		"http_host_is_required", "false",
+		tfw_cfg_set_bool,
+		&frang_cfg.http_host_is_required,
+	},
+	{
+		"http_ct_is_required", "false",
+		tfw_cfg_set_bool,
+		&frang_cfg.http_ct_is_required,
+	},
+	{
+		"http_methods", NULL,
+		frang_set_methods_mask,
+		.cleanup = frang_clear_methods_mask,
+	},
+	{
+		"http_ct_vals", NULL,
+		frang_set_ct_vals,
+		.cleanup = frang_free_ct_vals
+	},
+	{}
+};
+
+static TfwCfgSpec frang_cfg_toplevel_specs[] = {
+	{
+		"frang_limits", NULL,
+		tfw_cfg_handle_children,
+		&frang_cfg_section_specs
+	},
+	{}
+};
+
+static TfwCfgMod frang_cfg_mod = {
+	.name = "frang",
+	.specs = frang_cfg_toplevel_specs
+};
+
+static int __init
+frang_init(void)
+{
+	int r;
+
+	frang_mem_cache = KMEM_CACHE(frang_account_t, 0);
+	if (!frang_mem_cache) {
+		TFW_ERR("frang: can't create cache\n");
+		return -EINVAL;
+	}
+
+	r = tfw_cfg_mod_register(&frang_cfg_mod);
+	if (r) {
+		TFW_ERR("frang: can't register as a configuration module\n");
+		goto err_cfg;
+	}
+
+	r = tfw_classifier_register(&frang_class_ops);
+	if (r) {
+		TFW_ERR("frang: can't register classifier\n");
+		goto err_class;
+	}
+
+	/* FIXME: this is not a FSM here, but rather a set of static checks
+	 * that are executed when a HTTP request is fully parsed.
+	 * These checks should be executed during the parsing process in order
+	 * to drop suspicious requests as early as possible. */
+	r = tfw_gfsm_register_fsm(TFW_FSM_FRANG, frang_http_req_handler);
+	if (r) {
+		TFW_ERR("frang: can't register fsm: req\n");
+		goto err_fsm;
+	}
+
+	r = tfw_gfsm_register_hook(TFW_FSM_HTTP, TFW_GFSM_HOOK_PRIORITY_ANY,
+				   TFW_HTTP_FSM_REQ_MSG, 0,
+				   TFW_FSM_FRANG);
+	if (r) {
+		TFW_ERR("frang: can't register gfsm hook: req\n");
+		goto err_hook;
+	}
+
+	TFW_WARN("frang mudule can't be unloaded, "
+		 "so all allocated resources won't freed\n");
+
+	return 0;
+err_hook:
+	tfw_gfsm_unregister_fsm(TFW_FSM_FRANG);
+err_fsm:
+	tfw_classifier_unregister();
+err_class:
+	tfw_cfg_mod_unregister(&frang_cfg_mod);
+err_cfg:
+	kmem_cache_destroy(frang_mem_cache);
+	return r;
+}
+
+module_init(frang_init);
```

Comment:
Comment https://github.com/natsys/tempesta/pull/70#discussion_r26333673 is still not fully implemented.